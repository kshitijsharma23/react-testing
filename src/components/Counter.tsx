import React, { useState } from 'react';

import styles from './counterStyles.module.css';

enum CounterActions {
    INCREMENT = 'increment',
    DECREMENT = 'decrement',
}

const Counter: React.FC = () => {
    const [counter, setCounter] = useState(0);

    const handleCounter = (action: CounterActions) => (): void => {
        if (action === CounterActions.INCREMENT) {
            setCounter(counter => counter + 1);
        } else {
            setCounter(counter => counter - 1);
        }
    }

    const isCounterZero = counter === 0;

    return (
        <div className={styles['counter-container']}>
            <button
                disabled={isCounterZero}
                className="decrement"
                // onClick={handleCounter(CounterActions.DECREMENT)}
                onClick={isCounterZero ? undefined : handleCounter(CounterActions.DECREMENT)}
            >
                -
            </button>
            <span className="counter-value">{counter}</span>
            <button
                className="increment"
                onClick={handleCounter(CounterActions.INCREMENT)}
            >
                +
            </button>
        </div>
    );
}

export default Counter;
