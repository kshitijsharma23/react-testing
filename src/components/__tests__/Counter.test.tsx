import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

import Counter from '../Counter';

describe('Counter unit test cases', () => {
    let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

    beforeEach(() => {
        component = shallow(<Counter />);
    })

    it('should render snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('should have initial value as 0', () => {
        const counterText = component.find('span.counter-value').text();
        expect(counterText).toBe('0');
    });

    it('should disable decrement button when counter value is 0', () => {
        const decrementButton = component.find('button.decrement');
        const isDisabled = decrementButton.props().disabled;
        expect(isDisabled).toBeTruthy();
    });

    it('should not decrease counter when button is disabled', () => {
        const decrementButton = component.find('button.decrement');
        act(() => {
            decrementButton.simulate('click');
        });

        const counterText = component.find('span.counter-value').text();
        expect(counterText).toBe('0');
    });

    it('should increase counter on increment', () => {
        const incrementButton = component.find('button.increment');
        incrementButton.simulate('click');

        const counterText = component.find('span.counter-value').text();
        expect(counterText).toBe('1');
        expect(component).toMatchSnapshot();
    });

    xit('should decrease counter on decrement', () => {
        const decrementButton = component.find('button.decrement');
        decrementButton.simulate('click');

        const counterText = component.find('span.counter-value').text();
        expect(counterText).toBe('-1');
        expect(component).toMatchSnapshot();
    });

    it('should have correct value after multiple increments/decrements', () => {
        component.find('button.increment').simulate('click'); // 1
        expect(component.find('span.counter-value').text()).toBe('1');

        component.find('button.decrement').simulate('click'); // 0
        expect(component.find('span.counter-value').text()).toBe('0');

        component.find('button.increment').simulate('click'); // 1
        expect(component.find('span.counter-value').text()).toBe('1');

        component.find('button.decrement').simulate('click'); // 0
        expect(component.find('span.counter-value').text()).toBe('0');

        component.find('button.increment').simulate('click'); // 1
        expect(component.find('span.counter-value').text()).toBe('1');

        component.find('button.increment').simulate('click'); // 2
        expect(component.find('span.counter-value').text()).toBe('2');

        component.find('button.decrement').simulate('click'); // 0
        expect(component.find('span.counter-value').text()).toBe('1');

        component.find('button.decrement').simulate('click'); // 0
        expect(component.find('span.counter-value').text()).toBe('0');

        component.find('button.decrement').simulate('click'); // 0
        expect(component.find('span.counter-value').text()).toBe('0');
    });
});