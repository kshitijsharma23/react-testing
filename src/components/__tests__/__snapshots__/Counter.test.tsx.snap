// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter unit test cases should decrease counter on decrement 1`] = `ShallowWrapper {}`;

exports[`Counter unit test cases should increase counter on increment 1`] = `ShallowWrapper {}`;

exports[`Counter unit test cases should render snapshot 1`] = `ShallowWrapper {}`;
